mempool/src/shared_mempool/coordinator.rs

```diff
@@ -20,60 +23,90 @@ use channel::libra_channel;
 use debug_interface::prelude::*;
 use futures::{
     channel::{mpsc, oneshot},
-    stream::select_all,
+    stream::{select_all, FuturesUnordered},
     StreamExt,
 };
-use libra_config::config::NodeConfig;
+use libra_config::config::{NetworkId, NodeConfig, PeerNetworkId};
 use libra_logger::prelude::*;
 use libra_security_logger::{security_log, SecurityEvent};
-use libra_types::{on_chain_config::OnChainConfigPayload, transaction::SignedTransaction, PeerId};
+use libra_types::{on_chain_config::OnChainConfigPayload, transaction::SignedTransaction};
 use std::{
     ops::Deref,
     sync::{Arc, Mutex},
-    time::Duration,
+    time::{Duration, Instant},
 };
 use tokio::{runtime::Handle, time::interval};
 use vm_validator::vm_validator::TransactionValidation;
 
 /// Coordinator that handles [`SyncEvent`], which is periodically emitted for us to
 /// broadcast ready to go transactions to peers.
-pub(crate) async fn broadcast_coordinator<V>(smp: SharedMempool<V>, mut interval: IntervalStream)
-where
+pub(crate) async fn broadcast_coordinator<V>(
+    smp: SharedMempool<V>,
+    mut schedule_broadcast_rx: libra_channel::Receiver<PeerNetworkId, PeerNetworkId>,
+    executor: Handle,
+    // A stream that controls the broadcast loop. If provided, a single broadcast will happen when one event is emitted
+    // by `broadcast_ticker`. Should only be used for testing purposes to coordinate broadcast events
+    mut broadcast_ticker: Option<IntervalStream>,
+) where
     V: TransactionValidation,
 {
     let peer_manager = smp.peer_manager;
     let mempool = smp.mempool;
-    let network_senders = smp.network_senders;
+    let mut network_senders = smp.network_senders;
     let batch_size = smp.config.shared_mempool_batch_size;
-    let subscribers = smp.subscribers;
+    let mut scheduled_broadcasts = FuturesUnordered::new();
+    let interval_ms = smp.config.shared_mempool_tick_interval_ms;
+
+    loop {
+        tick(&mut broadcast_ticker).await;
+
+        ::futures::select! {
+            new_peer = schedule_broadcast_rx.select_next_some() => {
+                if tasks::broadcast_single_peer(new_peer, &peer_manager, &mempool, &mut network_senders, batch_size) {
```

Comment:
why not always schedule next broadcast and check on a fly if current one should be noop?
with current code, assuming primary is flaky, how fallback peer can be reliably scheduled ?